README.md
# Code of Kutulu

## Statement mistakes
- Shelters start with 10 energy . [..] Shelters get refilled to 10 every 50 turns.
  => Shelters starts with 0 energy. [...] Shelters get refilled to 10 every multiple of 50 turns, except turn 0.
- Slasher goes in RUSH mode at the end of SPAWN no matter what, instead of what is stated in the statement (RUSH if target or WANDER if not)

## Statement imprecision
- If a shelter is supposed to heal more than its energy for a turn (1 energy but two explorers for instance) everyone is healed and the shelter's energy is set to 0.
- PLAN and LIGHT are created by the explorer but they stay after the explorer's death until they fade out.
- Minions pick their new target AFTER explorers moves.
- The prefered direction URDL is offseted by 1 more each turn. Meaning a minion will preferer Up instead of Right at turn 0 but he preferers Right instead of Up at turn 1 (RDLU), and so on.
- Yells works once per player, per player (meaning each explorer can yell each other explorer)

## Known bugs

- PLAN still healing after death of a player (doesn't affect the score)
- If you stay on top of a slasher when he gots no previous target, he will be stunned (you are counted as 4 players alone)

# Other ressources

## Game design document
This is the first document we shared about the game
https://docs.google.com/document/d/1jQkYszU6UipM8QFGnKm3oSypcYnELiwpK0SM8uqgsfQ/edit?usp=sharing


codeOfKutulu/CHANGELOG.md
# What's new

## 1.0

## 0.9.1
* #132 - Optimisation - dont add ground on cell with walls
* #137 - rework tiles to remove seams
* #123 - center map title text
* Add the ability to choose a map in parameters (map=name of map)
* #145 - add parametrized constants values based on seed
* #148 - increase base life of wanderers to 30

## 0.9
* #133 - add tentacles to map border fora fancy effect
* #119 - Limit Pacman map to wood leagues only
* #129 - A wanderer in range of a light effect will not have a yellow cell.
* #131 - The tooltip for wanderer reads "'till death", should be "'till recall".
* #128 - Explorers and minions share entity IDs. 
* #125 - rename sanity/life to param0 + put it after x,y
* #115 - fix INVALID MOVE bug
* #117 - tint ground in dark green if kutulu is coming

## 0.1.9
* #113 - all shelters refill every 50 turns, even if not depleted
* #114 - kutulu deals 50 sanity loss when only one player left

## 0.1.8 
* #62 - animation for the stunned players
* remove minions' circle indicating their range as it is now 1
* fix timeout bug & invalid actions
* make bosses deterministic

## 0.1.7
* Change fear sprite (again)
* #106 - Tweaked problem statement (and resolved #106)
* #107 - default AI for the bronze league
* #108 - isolation icon not always shown in first turn
* #104 - fix plan that still emited light effect
* more mazes

## 0.1.6
* #94 - last frame is not shown (bump to sdk 2.1)
* #100 - bump map healing to 3 (* number of players in range) and decreased shelter to 5
* #99 - do real BFS effects for PLAN & LIGHT
* wood3 & wood2 bosses - 1st draft
* #96 - add a move command with pathfinding
* #98 - Change leagues'rules


## 0.1.5
* #93 - fix stub / entityType was coded as in.nextLine()
* #92 - remove swapindex & send player with real IDs but each player its bot first
* #91 - change index to id
* #64 - add a EFFECT_YELL in inputs to announce who are stuck (by who) param1 & param2 are used for caster & receiver

## 0.1.4
* #85 - remove extra leagues
* #85 - rework default AI
* rework animation of minions (with color mask)
* remove some unused images
* #83 - red pentragram on spawn
* #82 - cleanup statement
* #81 - fix issue32 bias selection
* #80 - fix stub linesize
* #79 - no shelter input in wood

## 0.1.3
* fix bug when 4 players used light (alpha > 1.0 causes an exception in SDK)
* #issue 62 - inverse updateMinions & spreadMadness to avoid "Minions ignore dying players"
* #issue 24 - lower initial sanity of players (to 250) and start slashers at 200
* bump to SDK 1.36 (fix missing graphics glitch when texture are not in browser cache)
* map : walls are now sent with char '@' (old was '#') to please cg clojure bots :)

## 0.1.2
* slasher behavior : still go to old target if still in Line of sight
* rework the texture (may cause bugs in PIXI on 1st load)
* fix topleft flame graphic glitch
* rework slashers behaviour (see github's issue #44)
* handle problems in parameter SEED
* fix end screen
* add eulerscheZahl to test class (for replay)
* fixbug - map tool
* fixed minion size

## 0.1.1
* new logo
* statement has been reworked
* entityType are now strings ("EXPLORER", "WANDERER", "SLASHER", "EFFECT_SHELTER", "EFFECT_PLAN", "EFFECT_LIGHT")
* change some sprites (tombs, solo wall)
* stuck players will not be able to yell (on second stucked turn)
* slasher will now update their target when rushing (issue #44) 

codeOfKutulu/src/main/java/codeofkutulu/Constants.java
package codeofkutulu;

import java.util.Random;

import com.codingame.game.Player;

import codeofkutulu.models.Vector2D;

public class Constants {

  public static final String FILE_SEPARATOR = System.getProperty("file.separator");

  public static final int MAX_PLAYERS = 4;
  public static final int MAX_TURNS = 200;

  public static final int MAX_DIRS = 4;
  public static final int UP = 0, RIGHT = 1, DOWN = 2, LEFT = 3, STAY = 4;
  public static final String[] directions = { "up", "right", "down", "left", "wait" };
  public static final Vector2D vdir[] = { Vector2D.UP, Vector2D.RIGHT, Vector2D.DOWN, Vector2D.LEFT };

  public static int SPREAD_MADNESS_PER_TURN_AMOUNT = 3; // [3..5]
  public static int SPREAD_MADNESS_PLAYER_PROX_PER_TURN_AMOUNT = 1; // [1..3]
  public static final int SPREAD_KUTULU_MADNESS_PER_TURN_AMOUNT = 50;
  public static final int IM_NOT_ALONE_RANGE = 2;

  public static final int PLAYER_TYPE = 0;
  public static final int WANDERER_TYPE = 1;
  public static final int SLASHER_TYPE = 2;
  public static final int EFFECT_SHELTER_TYPE = 3;
  public static final int EFFECT_PLAN_TYPE = 4;
  public static final int EFFECT_REPULSE_TYPE = 5;
  public static final int EFFECT_YELL_TYPE = 6;
  public static final String[] unittypes = { "EXPLORER", "WANDERER", "SLASHER", "EFFECT_SHELTER", "EFFECT_PLAN", "EFFECT_LIGHT", "EFFECT_YELL" };
  
  public static int WANDERER_SPAWN_TIME = 3; // [3..6]
  public static final int WANDERER_SPAWN_INTERVAL = 5;
  public static int WANDERER_LIFE_TIME = 40; // [20..40]
//  public static final int WANDERER_FRIGHTEN_RADIUS = 0;

  public static final int SLASHER_SPAWN_TIME = 6;
  public static final int SLASHER_RUSH_PREP = 2;
  public static final int SLASHER_RUSH_STUN = 6;
  public static final int SLASHER_SANITY_SPAWN = 200;

  public static final int MINION_SANITY_LOST = 20;
  public static final boolean MINION_SEARCH_CYCLE_DIRECTIONS = true;
  public static final boolean MINION_SEARCH_CYCLE_PLAYERS = true;

  public static final int PLAYER_MAX_SANITY = 250;
  public static final int PLAYER_MAX_PLAN_EFFECT = 2;
  public static final int PLAYER_MAX_LIGHT_EFFECT = 3;

  public static final int EFFECT_PLAN_MAX_DURATION = 5;
  public static final int EFFECT_PLAN_HEAL_RADIUS = 2;
  public static final int EFFECT_PLAN_HEAL = 3;

  public static final int SHELTER_SPAWN_INTERVAL = 50;
  public static final int EFFECT_MAX_SHELTER_ENERGY = 10;
  public static final int EFFECT_SHELTER_HEAL = 5;

  public static final int EFFECT_REPULSE_MAX_DURATION = 3;
  public static final int EFFECT_REPULSE_RADIUS = 5;
  public static final int EFFECT_REPULSE_WEIGHT = 4;

  // Visual constant
  public static final int CELL_SIZE = 64;
  public static final Vector2D SCREEN_DIM = new Vector2D(1920, 1080);

  public static final int PLAYER_INITIAL_STUCK_ROUNDS = 2;

  public static final int EFFECT_YELL_RADIUS = 1;


  public static Random random; // not so constant constant


  public static void calculateParametrizedConstants() {
    SPREAD_MADNESS_PER_TURN_AMOUNT =  randomBetween(3,6);;
    SPREAD_MADNESS_PLAYER_PROX_PER_TURN_AMOUNT = randomBetween(1,3);
    WANDERER_SPAWN_TIME = randomBetween(3, 6);
    WANDERER_LIFE_TIME = randomBetween(30,40);
  }

  private static int randomBetween(int min, int max) {
    if (max <= min) return max;
    return min + random.nextInt(max-min+1);
  }

  public static void sendParametrizedConstants(Player player) {
    StringBuilder sb = new StringBuilder();
    sb.append(SPREAD_MADNESS_PER_TURN_AMOUNT).append(" ");
    sb.append(SPREAD_MADNESS_PLAYER_PROX_PER_TURN_AMOUNT).append(" ");
    sb.append(WANDERER_SPAWN_TIME).append(" ");
    sb.append(WANDERER_LIFE_TIME).append(" ");
    
    player.sendInputLine(sb.toString().trim());
  }


}
